File: Src/CSharp/FTP-related/FolderViews/OpenFolderForm.cs
using System;
using System.Drawing;
using System.Collections;
using System.ComponentModel;
using System.Windows.Forms;
using Xceed.FileSystem;
using Xceed.Ftp;

namespace FolderViews
{
	/// <summary>
	/// Summary description for OpenFolderForm.
	/// </summary>
	public class OpenFolderForm : System.Windows.Forms.Form
	{
    private System.Windows.Forms.Label label1;
    private System.Windows.Forms.Button openButton;
    private System.Windows.Forms.Button cancelButton;
    private System.Windows.Forms.RadioButton drivesRadio;
    private System.Windows.Forms.RadioButton networkRadio;
    private System.Windows.Forms.RadioButton isolatedRadio;
    private System.Windows.Forms.Label networkLabel;
    private System.Windows.Forms.TextBox networkTextBox;
    private System.Windows.Forms.RadioButton memoryRadio;
    private System.Windows.Forms.RadioButton ftpRadio;
    private System.Windows.Forms.Label userLabel;
    private System.Windows.Forms.Label passwordLabel;
    private System.Windows.Forms.TextBox userTextBox;
    private System.Windows.Forms.TextBox passwordTextBox;
    private System.Windows.Forms.TextBox hostTextBox;
    private System.Windows.Forms.Label hostLabel;
    private System.Windows.Forms.TextBox portTextBox;
    private System.Windows.Forms.Label portLabel;
    private CheckBox useModeZCheckBox;
		/// <summary>
		/// Required designer variable.
		/// </summary>
		private System.ComponentModel.Container components = null;

		public OpenFolderForm()
		{
			//
			// Required for Windows Form Designer support
			//
			InitializeComponent();

			//
			// TODO: Add any constructor code after InitializeComponent call
			//
		}

		/// <summary>
		/// Clean up any resources being used.
		/// </summary>
		protected override void Dispose( bool disposing )
		{
			if( disposing )
			{
				if(components != null)
				{
					components.Dispose();
				}
			}
			base.Dispose( disposing );
		}

		#region Windows Form Designer generated code
		/// <summary>
		/// Required method for Designer support - do not modify
		/// the contents of this method with the code editor.
		/// </summary>
		private void InitializeComponent()
		{
      this.drivesRadio = new System.Windows.Forms.RadioButton();
      this.networkRadio = new System.Windows.Forms.RadioButton();
      this.isolatedRadio = new System.Windows.Forms.RadioButton();
      this.label1 = new System.Windows.Forms.Label();
      this.networkLabel = new System.Windows.Forms.Label();
      this.networkTextBox = new System.Windows.Forms.TextBox();
      this.memoryRadio = new System.Windows.Forms.RadioButton();
      this.ftpRadio = new System.Windows.Forms.RadioButton();
      this.userLabel = new System.Windows.Forms.Label();
      this.passwordLabel = new System.Windows.Forms.Label();
      this.userTextBox = new System.Windows.Forms.TextBox();
      this.passwordTextBox = new System.Windows.Forms.TextBox();
      this.hostTextBox = new System.Windows.Forms.TextBox();
      this.hostLabel = new System.Windows.Forms.Label();
      this.portTextBox = new System.Windows.Forms.TextBox();
      this.portLabel = new System.Windows.Forms.Label();
      this.openButton = new System.Windows.Forms.Button();
      this.cancelButton = new System.Windows.Forms.Button();
      this.useModeZCheckBox = new System.Windows.Forms.CheckBox();
      this.SuspendLayout();
      // 
      // drivesRadio
      // 
      this.drivesRadio.Location = new System.Drawing.Point( 16, 32 );
      this.drivesRadio.Name = "drivesRadio";
      this.drivesRadio.Size = new System.Drawing.Size( 168, 16 );
      this.drivesRadio.TabIndex = 0;
      this.drivesRadio.Text = "My computer\'s logical &drives";
      // 
      // networkRadio
      // 
      this.networkRadio.Location = new System.Drawing.Point( 16, 64 );
      this.networkRadio.Name = "networkRadio";
      this.networkRadio.Size = new System.Drawing.Size( 152, 16 );
      this.networkRadio.TabIndex = 1;
      this.networkRadio.Text = "A particular &network path";
      this.networkRadio.CheckedChanged += new System.EventHandler( this.networkRadio_CheckedChanged );
      // 
      // isolatedRadio
      // 
      this.isolatedRadio.Location = new System.Drawing.Point( 16, 144 );
      this.isolatedRadio.Name = "isolatedRadio";
      this.isolatedRadio.Size = new System.Drawing.Size( 200, 16 );
      this.isolatedRadio.TabIndex = 4;
      this.isolatedRadio.Text = "The current user\'s &isolated storage";
      // 
      // label1
      // 
      this.label1.Font = new System.Drawing.Font( "Microsoft Sans Serif", 8.25F, System.Drawing.FontStyle.Bold, System.Drawing.GraphicsUnit.Point, ( ( byte )( 0 ) ) );
      this.label1.Location = new System.Drawing.Point( 8, 8 );
      this.label1.Name = "label1";
      this.label1.Size = new System.Drawing.Size( 336, 16 );
      this.label1.TabIndex = 0;
      this.label1.Text = "Select folder to view:";
      // 
      // networkLabel
      // 
      this.networkLabel.Enabled = false;
      this.networkLabel.Location = new System.Drawing.Point( 48, 88 );
      this.networkLabel.Name = "networkLabel";
      this.networkLabel.Size = new System.Drawing.Size( 100, 16 );
      this.networkLabel.TabIndex = 2;
      this.networkLabel.Text = "Network path:";
      // 
      // networkTextBox
      // 
      this.networkTextBox.Enabled = false;
      this.networkTextBox.Location = new System.Drawing.Point( 48, 104 );
      this.networkTextBox.Name = "networkTextBox";
      this.networkTextBox.Size = new System.Drawing.Size( 296, 20 );
      this.networkTextBox.TabIndex = 3;
      // 
      // memoryRadio
      // 
      this.memoryRadio.Location = new System.Drawing.Point( 16, 176 );
      this.memoryRadio.Name = "memoryRadio";
      this.memoryRadio.Size = new System.Drawing.Size( 152, 16 );
      this.memoryRadio.TabIndex = 5;
      this.memoryRadio.Text = "A volatile &memory folder";
      // 
      // ftpRadio
      // 
      this.ftpRadio.Location = new System.Drawing.Point( 16, 208 );
      this.ftpRadio.Name = "ftpRadio";
      this.ftpRadio.Size = new System.Drawing.Size( 184, 16 );
      this.ftpRadio.TabIndex = 6;
      this.ftpRadio.Text = "An &FTP server\'s starting folder";
      this.ftpRadio.CheckedChanged += new System.EventHandler( this.ftpRadio_CheckedChanged );
      // 
      // userLabel
      // 
      this.userLabel.Enabled = false;
      this.userLabel.Location = new System.Drawing.Point( 48, 272 );
      this.userLabel.Name = "userLabel";
      this.userLabel.Size = new System.Drawing.Size( 112, 16 );
      this.userLabel.TabIndex = 12;
      this.userLabel.Text = "Username:";
      // 
      // passwordLabel
      // 
      this.passwordLabel.Enabled = false;
      this.passwordLabel.Location = new System.Drawing.Point( 168, 272 );
      this.passwordLabel.Name = "passwordLabel";
      this.passwordLabel.Size = new System.Drawing.Size( 104, 16 );
      this.passwordLabel.TabIndex = 14;
      this.passwordLabel.Text = "Password:";
      // 
      // userTextBox
      // 
      this.userTextBox.Enabled = false;
      this.userTextBox.Location = new System.Drawing.Point( 48, 288 );
      this.userTextBox.Name = "userTextBox";
      this.userTextBox.Size = new System.Drawing.Size( 112, 20 );
      this.userTextBox.TabIndex = 9;
      this.userTextBox.Text = "anonymous";
      // 
      // passwordTextBox
      // 
      this.passwordTextBox.Enabled = false;
      this.passwordTextBox.Location = new System.Drawing.Point( 168, 288 );
      this.passwordTextBox.Name = "passwordTextBox";
      this.passwordTextBox.Size = new System.Drawing.Size( 112, 20 );
      this.passwordTextBox.TabIndex = 10;
      this.passwordTextBox.Text = "guest";
      this.passwordTextBox.UseSystemPasswordChar = true;
      // 
      // hostTextBox
      // 
      this.hostTextBox.Enabled = false;
      this.hostTextBox.Location = new System.Drawing.Point( 48, 248 );
      this.hostTextBox.Name = "hostTextBox";
      this.hostTextBox.Size = new System.Drawing.Size( 176, 20 );
      this.hostTextBox.TabIndex = 7;
      this.hostTextBox.Text = "ftp.xceed.com";
      // 
      // hostLabel
      // 
      this.hostLabel.Enabled = false;
      this.hostLabel.Location = new System.Drawing.Point( 48, 232 );
      this.hostLabel.Name = "hostLabel";
      this.hostLabel.Size = new System.Drawing.Size( 176, 16 );
      this.hostLabel.TabIndex = 8;
      this.hostLabel.Text = "FTP server hostname or address:";
      // 
      // portTextBox
      // 
      this.portTextBox.Enabled = false;
      this.portTextBox.Location = new System.Drawing.Point( 232, 248 );
      this.portTextBox.Name = "portTextBox";
      this.portTextBox.Size = new System.Drawing.Size( 48, 20 );
      this.portTextBox.TabIndex = 8;
      this.portTextBox.Text = "21";
      // 
      // portLabel
      // 
      this.portLabel.Enabled = false;
      this.portLabel.Location = new System.Drawing.Point( 232, 232 );
      this.portLabel.Name = "portLabel";
      this.portLabel.Size = new System.Drawing.Size( 48, 16 );
      this.portLabel.TabIndex = 10;
      this.portLabel.Text = "Port:";
      // 
      // openButton
      // 
      this.openButton.DialogResult = System.Windows.Forms.DialogResult.OK;
      this.openButton.Location = new System.Drawing.Point( 189, 365 );
      this.openButton.Name = "openButton";
      this.openButton.Size = new System.Drawing.Size( 75, 23 );
      this.openButton.TabIndex = 12;
      this.openButton.Text = "&Open";
      // 
      // cancelButton
      // 
      this.cancelButton.DialogResult = System.Windows.Forms.DialogResult.Cancel;
      this.cancelButton.Location = new System.Drawing.Point( 269, 365 );
      this.cancelButton.Name = "cancelButton";
      this.cancelButton.Size = new System.Drawing.Size( 75, 23 );
      this.cancelButton.TabIndex = 13;
      this.cancelButton.Text = "&Cancel";
      // 
      // useModeZCheckBox
      // 
      this.useModeZCheckBox.AutoSize = true;
      this.useModeZCheckBox.Location = new System.Drawing.Point( 48, 325 );
      this.useModeZCheckBox.Name = "useModeZCheckBox";
      this.useModeZCheckBox.Size = new System.Drawing.Size( 85, 17 );
      this.useModeZCheckBox.TabIndex = 11;
      this.useModeZCheckBox.Text = "Use Mode Z";
      this.useModeZCheckBox.UseVisualStyleBackColor = true;
      this.useModeZCheckBox.CheckedChanged += new System.EventHandler( this.UseModeZCheckBox_CheckedChanged );
      // 
      // OpenFolderForm
      // 
      this.AcceptButton = this.openButton;
      this.AutoScaleBaseSize = new System.Drawing.Size( 5, 13 );
      this.CancelButton = this.cancelButton;
      this.ClientSize = new System.Drawing.Size( 360, 400 );
      this.ControlBox = false;
      this.Controls.Add( this.useModeZCheckBox );
      this.Controls.Add( this.cancelButton );
      this.Controls.Add( this.openButton );
      this.Controls.Add( this.portTextBox );
      this.Controls.Add( this.hostTextBox );
      this.Controls.Add( this.passwordTextBox );
      this.Controls.Add( this.userTextBox );
      this.Controls.Add( this.networkTextBox );
      this.Controls.Add( this.portLabel );
      this.Controls.Add( this.hostLabel );
      this.Controls.Add( this.passwordLabel );
      this.Controls.Add( this.userLabel );
      this.Controls.Add( this.ftpRadio );
      this.Controls.Add( this.memoryRadio );
      this.Controls.Add( this.networkLabel );
      this.Controls.Add( this.label1 );
      this.Controls.Add( this.isolatedRadio );
      this.Controls.Add( this.networkRadio );
      this.Controls.Add( this.drivesRadio );
      this.FormBorderStyle = System.Windows.Forms.FormBorderStyle.FixedDialog;
      this.MaximizeBox = false;
      this.MinimizeBox = false;
      this.Name = "OpenFolderForm";
      this.StartPosition = System.Windows.Forms.FormStartPosition.CenterParent;
      this.Text = "Open a new folder";
      this.Load += new System.EventHandler( this.OpenFolderForm_Load );
      this.ResumeLayout( false );
      this.PerformLayout();

    }
		#endregion

    public new FolderForm ShowDialog( IWin32Window owner )
    {
      if( base.ShowDialog( owner ) == DialogResult.OK )
      {
        AbstractFolder[] folders = null;
        string title = string.Empty;

        if( drivesRadio.Checked )
        {
          string[] drives = System.IO.Directory.GetLogicalDrives();
          folders = new AbstractFolder[ drives.Length ];

          for( int index=0; index<drives.Length; index++ )
          {
            try
            {
              folders[ index ] = new DiskFolder( drives[ index ] );
            }
            catch {}
          }

          title = "My Computer's logical drives";
        }
        else if( networkRadio.Checked )
        {
          try
          {
            folders = new AbstractFolder[] { new DiskFolder( networkTextBox.Text ) };
          }
          catch( Exception except )
          {
            MessageBox.Show( 
              "The path \"" + networkTextBox.Text + "\" could not be found.\n" + except.Message,
              "Error",
              MessageBoxButtons.OK, 
              MessageBoxIcon.Error );

            return null;
          }

          title = "Path: " + networkTextBox.Text;
        }
        else if( isolatedRadio.Checked )
        {
          try
          {
            folders = new AbstractFolder[] { new IsolatedFolder( @"\" ) };
          }
          catch( Exception except )
          {
            MessageBox.Show( 
              "There was an error accessing the isolated storage.\n" + except.Message,
              "Error",
              MessageBoxButtons.OK, 
              MessageBoxIcon.Error );

            return null;
          }

          title = "Isolated storage";
        }
        else if( memoryRadio.Checked )
        {
          folders = new AbstractFolder[] { new MemoryFolder() };
          title = "Memory folder: " + folders[ 0 ].FullName;
        }
        else if( ftpRadio.Checked )
        {
          try
          {
            FtpConnection connection = new FtpConnection( 
              hostTextBox.Text, 
              int.Parse( portTextBox.Text ),
              userTextBox.Text,
              passwordTextBox.Text );

            if( useModeZCheckBox.Checked )
            {
              connection.TransferMode = TransferMode.ZLibCompressed;
            }

            folders = new AbstractFolder[] { new FtpFolder( connection ) };
          }
          catch( Exception except )
          {
            MessageBox.Show( 
              "There was an error connecting to the FTP server.\n" + except.Message,
              "Error",
              MessageBoxButtons.OK, 
              MessageBoxIcon.Error );

            return null;
          }

          title = "FTP server: " + hostTextBox.Text;
        }

        if( folders == null )
          return null;

        // As you can see, the same FolderForm type is used to display all these
        // kinds of AbstractFolder instances.
        return new FolderForm( title, folders );
      }

      return null;
    }

    private void networkRadio_CheckedChanged(object sender, System.EventArgs e)
    {
      networkLabel.Enabled = networkRadio.Checked;
      networkTextBox.Enabled = networkRadio.Checked;
    }

    private void ftpRadio_CheckedChanged(object sender, System.EventArgs e)
    {
      hostLabel.Enabled = ftpRadio.Checked;
      hostTextBox.Enabled = ftpRadio.Checked;
      portLabel.Enabled = ftpRadio.Checked;
      portTextBox.Enabled = ftpRadio.Checked;
      userLabel.Enabled = ftpRadio.Checked;
      userTextBox.Enabled = ftpRadio.Checked;
      passwordLabel.Enabled = ftpRadio.Checked;
      passwordTextBox.Enabled = ftpRadio.Checked;
      useModeZCheckBox.Enabled = ftpRadio.Checked;
    }

    private void OpenFolderForm_Load(object sender, System.EventArgs e)
    {
      ftpRadio.Checked = true;
    }

    private void UseModeZCheckBox_CheckedChanged( object sender, EventArgs e )
    {
      
    }
	}
}

File: Src/Visual Basic .NET/FTP-related/FolderViews/OpenFolderForm.vb
Imports System
Imports System.Drawing
Imports System.Collections
Imports System.ComponentModel
Imports System.Windows.Forms
Imports Xceed.FileSystem
Imports Xceed.Ftp

Namespace FolderViews
    Public Class OpenFolderForm
        Inherits System.Windows.Forms.Form

        Private label1 As System.Windows.Forms.Label
        Private openButton As System.Windows.Forms.Button
        Private Shadows cancelButton As System.Windows.Forms.Button
        Private drivesRadio As System.Windows.Forms.RadioButton
        WithEvents networkRadio As System.Windows.Forms.RadioButton
        Private isolatedRadio As System.Windows.Forms.RadioButton
        Private networkLabel As System.Windows.Forms.Label
        Private networkTextBox As System.Windows.Forms.TextBox
        Private memoryRadio As System.Windows.Forms.RadioButton
        WithEvents ftpRadio As System.Windows.Forms.RadioButton
        Private userLabel As System.Windows.Forms.Label
        Private passwordLabel As System.Windows.Forms.Label
        Private userTextBox As System.Windows.Forms.TextBox
        Private passwordTextBox As System.Windows.Forms.TextBox
        Private hostTextBox As System.Windows.Forms.TextBox
        Private hostLabel As System.Windows.Forms.Label
        Private portTextBox As System.Windows.Forms.TextBox
        Private portLabel As System.Windows.Forms.Label
        WithEvents useModeZCheckBox As System.Windows.Forms.CheckBox

        Private components As System.ComponentModel.Container = Nothing

        Public Sub New()
            InitializeComponent()
        End Sub

        Protected Overloads Overrides Sub Dispose(ByVal disposing As Boolean)
            If (disposing) Then
                If Not components Is Nothing Then
                    components.Dispose()
                End If
            End If
            MyBase.Dispose(disposing)
        End Sub

#Region "Windows Form Designer generated code"
        Private Sub InitializeComponent()
            Me.drivesRadio = New System.Windows.Forms.RadioButton
            Me.networkRadio = New System.Windows.Forms.RadioButton
            Me.isolatedRadio = New System.Windows.Forms.RadioButton
            Me.label1 = New System.Windows.Forms.Label
            Me.networkLabel = New System.Windows.Forms.Label
            Me.networkTextBox = New System.Windows.Forms.TextBox
            Me.memoryRadio = New System.Windows.Forms.RadioButton
            Me.ftpRadio = New System.Windows.Forms.RadioButton
            Me.userLabel = New System.Windows.Forms.Label
            Me.passwordLabel = New System.Windows.Forms.Label
            Me.userTextBox = New System.Windows.Forms.TextBox
            Me.passwordTextBox = New System.Windows.Forms.TextBox
            Me.hostTextBox = New System.Windows.Forms.TextBox
            Me.hostLabel = New System.Windows.Forms.Label
            Me.portTextBox = New System.Windows.Forms.TextBox
            Me.portLabel = New System.Windows.Forms.Label
            Me.openButton = New System.Windows.Forms.Button
            Me.cancelButton = New System.Windows.Forms.Button
            Me.useModeZCheckBox = New System.Windows.Forms.CheckBox
            Me.SuspendLayout()
            '
            'drivesRadio
            '
            Me.drivesRadio.Location = New System.Drawing.Point(16, 32)
            Me.drivesRadio.Name = "drivesRadio"
            Me.drivesRadio.Size = New System.Drawing.Size(168, 16)
            Me.drivesRadio.TabIndex = 0
            Me.drivesRadio.Text = "My computer's logical &drives"
            '
            'networkRadio
            '
            Me.networkRadio.Location = New System.Drawing.Point(16, 64)
            Me.networkRadio.Name = "networkRadio"
            Me.networkRadio.Size = New System.Drawing.Size(152, 16)
            Me.networkRadio.TabIndex = 1
            Me.networkRadio.Text = "A particular &network path"
            '
            'isolatedRadio
            '
            Me.isolatedRadio.Location = New System.Drawing.Point(16, 144)
            Me.isolatedRadio.Name = "isolatedRadio"
            Me.isolatedRadio.Size = New System.Drawing.Size(200, 16)
            Me.isolatedRadio.TabIndex = 3
            Me.isolatedRadio.Text = "The current user's &isolated storage"
            '
            'label1
            '
            Me.label1.Font = New System.Drawing.Font("Microsoft Sans Serif", 8.25!, System.Drawing.FontStyle.Bold, System.Drawing.GraphicsUnit.Point, CType(0, Byte))
            Me.label1.Location = New System.Drawing.Point(8, 8)
            Me.label1.Name = "label1"
            Me.label1.Size = New System.Drawing.Size(336, 16)
            Me.label1.TabIndex = 0
            Me.label1.Text = "Select folder to view:"
            '
            'networkLabel
            '
            Me.networkLabel.Enabled = False
            Me.networkLabel.Location = New System.Drawing.Point(48, 88)
            Me.networkLabel.Name = "networkLabel"
            Me.networkLabel.Size = New System.Drawing.Size(100, 16)
            Me.networkLabel.TabIndex = 3
            Me.networkLabel.Text = "Network path:"
            '
            'networkTextBox
            '
            Me.networkTextBox.Enabled = False
            Me.networkTextBox.Location = New System.Drawing.Point(48, 104)
            Me.networkTextBox.Name = "networkTextBox"
            Me.networkTextBox.Size = New System.Drawing.Size(296, 20)
            Me.networkTextBox.TabIndex = 2
            '
            'memoryRadio
            '
            Me.memoryRadio.Location = New System.Drawing.Point(16, 176)
            Me.memoryRadio.Name = "memoryRadio"
            Me.memoryRadio.Size = New System.Drawing.Size(152, 16)
            Me.memoryRadio.TabIndex = 4
            Me.memoryRadio.Text = "A volatile &memory folder"
            '
            'ftpRadio
            '
            Me.ftpRadio.Location = New System.Drawing.Point(16, 208)
            Me.ftpRadio.Name = "ftpRadio"
            Me.ftpRadio.Size = New System.Drawing.Size(184, 16)
            Me.ftpRadio.TabIndex = 5
            Me.ftpRadio.Text = "An &FTP server's starting folder"
            '
            'userLabel
            '
            Me.userLabel.Enabled = False
            Me.userLabel.Location = New System.Drawing.Point(48, 272)
            Me.userLabel.Name = "userLabel"
            Me.userLabel.Size = New System.Drawing.Size(112, 16)
            Me.userLabel.TabIndex = 12
            Me.userLabel.Text = "Username:"
            '
            'passwordLabel
            '
            Me.passwordLabel.Enabled = False
            Me.passwordLabel.Location = New System.Drawing.Point(168, 272)
            Me.passwordLabel.Name = "passwordLabel"
            Me.passwordLabel.Size = New System.Drawing.Size(104, 16)
            Me.passwordLabel.TabIndex = 14
            Me.passwordLabel.Text = "Password:"
            '
            'userTextBox
            '
            Me.userTextBox.Enabled = False
            Me.userTextBox.Location = New System.Drawing.Point(48, 288)
            Me.userTextBox.Name = "userTextBox"
            Me.userTextBox.Size = New System.Drawing.Size(112, 20)
            Me.userTextBox.TabIndex = 8
            Me.userTextBox.Text = "anonymous"
            '
            'passwordTextBox
            '
            Me.passwordTextBox.Enabled = False
            Me.passwordTextBox.Location = New System.Drawing.Point(168, 288)
            Me.passwordTextBox.Name = "passwordTextBox"
            Me.passwordTextBox.Size = New System.Drawing.Size(112, 20)
            Me.passwordTextBox.TabIndex = 9
            Me.passwordTextBox.Text = "guest"
            Me.passwordTextBox.UseSystemPasswordChar = True
            '
            'hostTextBox
            '
            Me.hostTextBox.Enabled = False
            Me.hostTextBox.Location = New System.Drawing.Point(48, 248)
            Me.hostTextBox.Name = "hostTextBox"
            Me.hostTextBox.Size = New System.Drawing.Size(176, 20)
            Me.hostTextBox.TabIndex = 6
            Me.hostTextBox.Text = "ftp.xceed.com"
            '
            'hostLabel
            '
            Me.hostLabel.Enabled = False
            Me.hostLabel.Location = New System.Drawing.Point(48, 232)
            Me.hostLabel.Name = "hostLabel"
            Me.hostLabel.Size = New System.Drawing.Size(176, 16)
            Me.hostLabel.TabIndex = 8
            Me.hostLabel.Text = "FTP server hostname or address:"
            '
            'portTextBox
            '
            Me.portTextBox.Enabled = False
            Me.portTextBox.Location = New System.Drawing.Point(232, 248)
            Me.portTextBox.Name = "portTextBox"
            Me.portTextBox.Size = New System.Drawing.Size(48, 20)
            Me.portTextBox.TabIndex = 7
            Me.portTextBox.Text = "21"
            '
            'portLabel
            '
            Me.portLabel.Enabled = False
            Me.portLabel.Location = New System.Drawing.Point(232, 232)
            Me.portLabel.Name = "portLabel"
            Me.portLabel.Size = New System.Drawing.Size(48, 16)
            Me.portLabel.TabIndex = 10
            Me.portLabel.Text = "Port:"
            '
            'openButton
            '
            Me.openButton.DialogResult = System.Windows.Forms.DialogResult.OK
            Me.openButton.Location = New System.Drawing.Point(189, 364)
            Me.openButton.Name = "openButton"
            Me.openButton.Size = New System.Drawing.Size(75, 23)
            Me.openButton.TabIndex = 11
            Me.openButton.Text = "&Open"
            '
            'cancelButton
            '
            Me.cancelButton.DialogResult = System.Windows.Forms.DialogResult.Cancel
            Me.cancelButton.Location = New System.Drawing.Point(269, 364)
            Me.cancelButton.Name = "cancelButton"
            Me.cancelButton.Size = New System.Drawing.Size(75, 23)
            Me.cancelButton.TabIndex = 12
            Me.cancelButton.Text = "&Cancel"
            '
            'useModeZCheckBox
            '
            Me.useModeZCheckBox.AutoSize = True
            Me.useModeZCheckBox.Location = New System.Drawing.Point(48, 324)
            Me.useModeZCheckBox.Name = "useModeZCheckBox"
            Me.useModeZCheckBox.Size = New System.Drawing.Size(85, 17)
            Me.useModeZCheckBox.TabIndex = 10
            Me.useModeZCheckBox.Text = "Use Mode Z"
            Me.useModeZCheckBox.UseVisualStyleBackColor = True
            '
            'OpenFolderForm
            '
            Me.AcceptButton = Me.openButton
            Me.AutoScaleBaseSize = New System.Drawing.Size(5, 13)
            Me.ClientSize = New System.Drawing.Size(360, 399)
            Me.ControlBox = False
            Me.Controls.Add(Me.useModeZCheckBox)
            Me.Controls.Add(Me.cancelButton)
            Me.Controls.Add(Me.openButton)
            Me.Controls.Add(Me.portTextBox)
            Me.Controls.Add(Me.hostTextBox)
            Me.Controls.Add(Me.passwordTextBox)
            Me.Controls.Add(Me.userTextBox)
            Me.Controls.Add(Me.networkTextBox)
            Me.Controls.Add(Me.portLabel)
            Me.Controls.Add(Me.hostLabel)
            Me.Controls.Add(Me.passwordLabel)
            Me.Controls.Add(Me.userLabel)
            Me.Controls.Add(Me.ftpRadio)
            Me.Controls.Add(Me.memoryRadio)
            Me.Controls.Add(Me.networkLabel)
            Me.Controls.Add(Me.label1)
            Me.Controls.Add(Me.isolatedRadio)
            Me.Controls.Add(Me.networkRadio)
            Me.Controls.Add(Me.drivesRadio)
            Me.FormBorderStyle = System.Windows.Forms.FormBorderStyle.FixedDialog
            Me.MaximizeBox = False
            Me.MinimizeBox = False
            Me.Name = "OpenFolderForm"
            Me.StartPosition = System.Windows.Forms.FormStartPosition.CenterParent
            Me.Text = "Open a new folder"
            Me.ResumeLayout(False)
            Me.PerformLayout()

        End Sub
#End Region

        Public Shadows Function ShowDialog(ByVal owner As IWin32Window) As FolderForm
            If MyBase.ShowDialog(owner) = System.Windows.Forms.DialogResult.OK Then
                Dim folders As AbstractFolder() = Nothing

                Dim title As String = String.Empty

                If (drivesRadio.Checked) Then
                    Dim drives As String() = System.IO.Directory.GetLogicalDrives()
                    folders = New AbstractFolder(drives.Length - 1) {}

                    Dim index As Integer
                    For index = 0 To drives.Length - 1
                        Try
                            folders(index) = New DiskFolder(drives(index))
                        Catch
                        End Try
                    Next index

                    title = "My Computer's logical drives"
                ElseIf (networkRadio.Checked) Then
                    Try
                        folders = New AbstractFolder() {New DiskFolder(networkTextBox.Text)}
                    Catch except As Exception
                        MessageBox.Show("The path " + networkTextBox.Text + " could not be found." + Environment.NewLine + except.Message, "Error", MessageBoxButtons.OK, MessageBoxIcon.Error)

                        Return Nothing
                    End Try
                    title = "Path: " + networkTextBox.Text

                ElseIf (isolatedRadio.Checked) Then
                    Try
                        folders = New AbstractFolder() {New IsolatedFolder("\")}
                    Catch except As Exception
                        MessageBox.Show("There was an error accessing the isolated storage.\n" + except.Message, "Error", MessageBoxButtons.OK, MessageBoxIcon.Error)

                        Return Nothing
                    End Try

                    title = "Isolated storage"
                ElseIf (memoryRadio.Checked) Then
                    folders = New AbstractFolder() {New MemoryFolder}
                    title = "Memory folder: " + folders(0).FullName
                ElseIf (ftpRadio.Checked) Then
                    Try
                        Dim connection As New FtpConnection(hostTextBox.Text, Integer.Parse(portTextBox.Text), userTextBox.Text, passwordTextBox.Text)

                        If (useModeZCheckBox.Checked) Then
                            connection.TransferMode = TransferMode.ZLibCompressed
                        End If

                        folders = New AbstractFolder() {New FtpFolder(connection)}
                    Catch except As Exception
                        MessageBox.Show("There was an error connecting to the FTP server.\n" + except.Message, "Error", MessageBoxButtons.OK, MessageBoxIcon.Error)

                        Return Nothing
                    End Try

                    title = "FTP server: " + hostTextBox.Text
                End If

                If folders Is Nothing Then
                    Return Nothing
                End If

                ' As you can see, the same FolderForm type is used to display all these
                ' kinds of AbstractFolder instances.
                Return New FolderForm(title, folders)
            End If

            Return Nothing
        End Function

        Private Sub networkRadio_CheckedChanged(ByVal sender As Object, ByVal e As System.EventArgs) Handles networkRadio.CheckedChanged
            networkLabel.Enabled = networkRadio.Checked
            networkTextBox.Enabled = networkRadio.Checked
        End Sub

        Private Sub ftpRadio_CheckedChanged(ByVal sender As Object, ByVal e As System.EventArgs) Handles ftpRadio.CheckedChanged
            hostLabel.Enabled = ftpRadio.Checked
            hostTextBox.Enabled = ftpRadio.Checked
            portLabel.Enabled = ftpRadio.Checked
            portTextBox.Enabled = ftpRadio.Checked
            userLabel.Enabled = ftpRadio.Checked
            userTextBox.Enabled = ftpRadio.Checked
            passwordLabel.Enabled = ftpRadio.Checked
            passwordTextBox.Enabled = ftpRadio.Checked
            useModeZCheckBox.Enabled = ftpRadio.Checked
        End Sub

        Private Sub OpenFolderForm_Load(ByVal sender As Object, ByVal e As System.EventArgs) Handles MyBase.Load
            ftpRadio.Checked = True
        End Sub
    End Class
End Namespace